File: crates/geoflow-core/src/render.rs
//! Diagnostic renderers: plain text, JSON, JUnit XML.

use crate::diagnostics::{Diagnostic, Severity};

/// Output format for diagnostic rendering.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Format {
    Text,
    Json,
    Junit,
}

impl Format {
    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "text" => Some(Format::Text),
            "json" => Some(Format::Json),
            "junit" => Some(Format::Junit),
            _ => None,
        }
    }
}

/// Render diagnostics in the requested format.
pub fn render(diagnostics: &[Diagnostic], format: Format) -> String {
    match format {
        Format::Text => render_text(diagnostics),
        Format::Json => render_json(diagnostics),
        Format::Junit => render_junit(diagnostics),
    }
}

fn render_text(d: &[Diagnostic]) -> String {
    let mut out = String::new();
    for diag in d {
        let loc = format_location(diag);
        let fix = diag
            .fix_id
            .as_deref()
            .map(|f| format!(" [auto-fixable: {f}]"))
            .unwrap_or_default();
        out.push_str(&format!(
            "{sev}: [{id}] {msg}{loc}{fix}\n",
            sev = diag.severity,
            id = diag.rule_id,
            msg = diag.message,
            loc = loc,
            fix = fix,
        ));
    }
    let (e, w, i) = counts(d);
    out.push_str(&format!("\nsummary: {e} error, {w} warning, {i} info\n"));
    out
}

fn format_location(d: &Diagnostic) -> String {
    let mut parts = Vec::new();
    if let Some(file) = &d.location.file {
        parts.push(format!("file={file}"));
    }
    if let Some(group) = &d.location.group {
        parts.push(format!("group={group}"));
    }
    if let Some(line) = d.location.line {
        parts.push(format!("line={line}"));
    }
    if parts.is_empty() {
        String::new()
    } else {
        format!(" ({})", parts.join(", "))
    }
}

fn render_json(d: &[Diagnostic]) -> String {
    serde_json::to_string_pretty(d).expect("Diagnostic always serializes")
}

fn render_junit(d: &[Diagnostic]) -> String {
    let (e, w, _i) = counts(d);
    let mut out = String::new();
    out.push_str(r#"<?xml version="1.0" encoding="UTF-8"?>"#);
    out.push('\n');
    out.push_str(&format!(
        r#"<testsuite name="geoflow" tests="{}" failures="{}" errors="0" skipped="{}">"#,
        d.len(),
        e,
        w,
    ));
    out.push('\n');
    for diag in d {
        let name = xml_escape(&diag.rule_id);
        let msg = xml_escape(&diag.message);
        let class = xml_escape(diag.location.group.as_deref().unwrap_or("file"));
        match diag.severity {
            Severity::Error => {
                out.push_str(&format!(
                    "  <testcase classname=\"{class}\" name=\"{name}\">\n    <failure message=\"{msg}\"/>\n  </testcase>\n"
                ));
            }
            Severity::Warning => {
                out.push_str(&format!(
                    "  <testcase classname=\"{class}\" name=\"{name}\">\n    <skipped message=\"{msg}\"/>\n  </testcase>\n"
                ));
            }
            Severity::Info => {
                out.push_str(&format!(
                    "  <testcase classname=\"{class}\" name=\"{name}\">\n    <system-out>{msg}</system-out>\n  </testcase>\n"
                ));
            }
        }
    }
    out.push_str("</testsuite>\n");
    out
}

fn xml_escape(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        match c {
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '&' => out.push_str("&amp;"),
            '"' => out.push_str("&quot;"),
            '\'' => out.push_str("&apos;"),
            _ => out.push(c),
        }
    }
    out
}

fn counts(d: &[Diagnostic]) -> (usize, usize, usize) {
    let mut e = 0;
    let mut w = 0;
    let mut i = 0;
    for diag in d {
        match diag.severity {
            Severity::Error => e += 1,
            Severity::Warning => w += 1,
            Severity::Info => i += 1,
        }
    }
    (e, w, i)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> Vec<Diagnostic> {
        vec![
            Diagnostic::new("R1", Severity::Error, "boom").at_group("LOCA"),
            Diagnostic::new("R2", Severity::Warning, "watch out").at_line(3),
        ]
    }

    #[test]
    fn text_renders_summary() {
        let s = render(&sample(), Format::Text);
        assert!(s.contains("error: [R1] boom"));
        assert!(s.contains("warning: [R2] watch out"));
        assert!(s.contains("1 error, 1 warning"));
    }

    #[test]
    fn json_round_trips() {
        let s = render(&sample(), Format::Json);
        let back: Vec<Diagnostic> = serde_json::from_str(&s).unwrap();
        assert_eq!(back, sample());
    }

    #[test]
    fn junit_well_formed() {
        let s = render(&sample(), Format::Junit);
        assert!(s.contains("<?xml"));
        assert!(s.contains("<testsuite"));
        assert!(s.contains("<failure"));
        assert!(s.contains("<skipped"));
    }

    #[test]
    fn xml_escape_handles_special() {
        assert_eq!(xml_escape("a<b&c>\"'"), "a&lt;b&amp;c&gt;&quot;&apos;");
    }

    #[test]
    fn parse_format_strings() {
        assert_eq!(Format::parse("text"), Some(Format::Text));
        assert_eq!(Format::parse("json"), Some(Format::Json));
        assert_eq!(Format::parse("junit"), Some(Format::Junit));
        assert_eq!(Format::parse("yaml"), None);
    }
}

File: packages/core/src/render.ts
import { Option } from "effect";
import { Diagnostic, Severity } from "./diagnostics.js";

export enum Format {
  Text = "text",
  Json = "json",
  Junit = "junit",
}

export namespace Format {
  export function parse(value: string): Format | null {
    switch (value) {
      case "text":
        return Format.Text;
      case "json":
        return Format.Json;
      case "junit":
        return Format.Junit;
      default:
        return null;
    }
  }
}

export function render(diagnostics: readonly Diagnostic[], format: Format): string {
  switch (format) {
    case Format.Text:
      return renderText(diagnostics);
    case Format.Json:
      return renderJson(diagnostics);
    case Format.Junit:
      return renderJunit(diagnostics);
  }
}

function renderText(diagnostics: readonly Diagnostic[]): string {
  let output = "";
  for (const diagnostic of diagnostics) {
    const location = formatLocation(diagnostic);
    const fix = Option.match(diagnostic.fix_id, {
      onNone: () => "",
      onSome: (value) => ` [auto-fixable: ${value}]`,
    });
    output += `${diagnostic.severity}: [${diagnostic.rule_id}] ${diagnostic.message}${location}${fix}\n`;
  }

  const counts = countSeverities(diagnostics);
  output += `\nsummary: ${counts.error} error, ${counts.warning} warning, ${counts.info} info\n`;
  return output;
}

function formatLocation(diagnostic: Diagnostic): string {
  const parts: string[] = [];
  Option.match(diagnostic.location.file, {
    onNone: () => undefined,
    onSome: (value) => parts.push(`file=${value}`),
  });
  Option.match(diagnostic.location.group, {
    onNone: () => undefined,
    onSome: (value) => parts.push(`group=${value}`),
  });
  Option.match(diagnostic.location.line, {
    onNone: () => undefined,
    onSome: (value) => parts.push(`line=${value}`),
  });

  return parts.length === 0 ? "" : ` (${parts.join(", ")})`;
}

function renderJson(diagnostics: readonly Diagnostic[]): string {
  const normalized = diagnostics.map((diagnostic) => {
    const base = {
      rule_id: diagnostic.rule_id,
      severity: diagnostic.severity,
      message: diagnostic.message,
      location: {
        file: optionToJsonValue(diagnostic.location.file),
        line: optionToJsonValue(diagnostic.location.line),
        column: optionToJsonValue(diagnostic.location.column),
        group: optionToJsonValue(diagnostic.location.group),
        row_index: optionToJsonValue(diagnostic.location.row_index),
      },
    } as Record<string, unknown>;

    if (diagnostic.fix_id._tag === "Some") {
      base.fix_id = diagnostic.fix_id.value;
    }

    return base;
  });

  return `${JSON.stringify(normalized, null, 2)}\n`;
}

function optionToJsonValue(value: { _tag: string; value?: unknown }): unknown {
  return value._tag === "Some" ? (value.value ?? null) : null;
}

function renderJunit(diagnostics: readonly Diagnostic[]): string {
  const counts = countSeverities(diagnostics);
  let output = '<?xml version="1.0" encoding="UTF-8"?>\n';
  output += `<testsuite name="geoflow" tests="${diagnostics.length}" failures="${counts.error}" errors="0" skipped="${counts.warning}">\n`;

  for (const diagnostic of diagnostics) {
    const name = xmlEscape(diagnostic.rule_id);
    const message = xmlEscape(diagnostic.message);
    const className = xmlEscape(
      Option.getOrElse(diagnostic.location.group, () => "file")
    );

    switch (diagnostic.severity) {
      case Severity.Error:
        output += `  <testcase classname="${className}" name="${name}">\n`;
        output += `    <failure message="${message}"/>\n`;
        output += "  </testcase>\n";
        break;
      case Severity.Warning:
        output += `  <testcase classname="${className}" name="${name}">\n`;
        output += `    <skipped message="${message}"/>\n`;
        output += "  </testcase>\n";
        break;
      case Severity.Info:
        output += `  <testcase classname="${className}" name="${name}">\n`;
        output += `    <system-out>${message}</system-out>\n`;
        output += "  </testcase>\n";
        break;
    }
  }

  output += "</testsuite>\n";
  return output;
}

function xmlEscape(value: string): string {
  return value
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&apos;");
}

function countSeverities(diagnostics: readonly Diagnostic[]) {
  let error = 0;
  let warning = 0;
  let info = 0;

  for (const diagnostic of diagnostics) {
    switch (diagnostic.severity) {
      case Severity.Error:
        error += 1;
        break;
      case Severity.Warning:
        warning += 1;
        break;
      case Severity.Info:
        info += 1;
        break;
    }
  }

  return { error, warning, info };
}
